### DISCUSSION TITLE
How did you do the blueprintjs integration?

### DISCUSSION THREAD

**Discussion by HoffiMuc** [joffrey-bion/seven-wonders#75] [Q&A]
Hi, I am a backend developer doing my first steps with frontend (still loving typesafe kotlinJS now in multiplatform projects). As I am playing around with tutorials and ramp up my knowledge (on how to integrate/import external npm packages), I just copied your `sw-ui/src/main/kotlin/com/palantir/blueprintjs/` package contents into my project and it works like a charm! (I am jealous).

As I tried to e.g. integrate https://github.com/STRML/react-grid-layout and failed to succeed, I ask myself "how did you come up with the code that wraps blueprintjs to be usable via kotlinJS?

How did you "know" which things to "extract" to kotlin from the javascript node module?

Any tips/recommendations/hints/pointers for me on how to do so?

Thanks in advance and I appologize for the off-topic question here ...

**Comment by joffrey-bion** [comment]
Hi!

Usually if you're lucky you can use the tool Dukat which helps converting Typescript definitions to Kotlin external declarations.

In my case with blueprintjs (and I think with all react component libraries in general) Dukat struggles a bit, and a lot of manual work is necessary to tweak the Kotlin declarations.

In the end I ended up looking at the Typescript definitions myself and converting manually some of them, but the Kotlin/JS docs in this respect are pretty scarce.

You can find the path to the Typescript declarations of a library in the package.json file of the lib in general (the "typings" field). And it gives you the root of the Typescript files.

After converting these declarations manually, I also declared extensions on `RBuilder` to make it more convenient to use. That's also manual work.

The best would be to create a library out of these declarations, but I didn't take the time to do so yet. Will do though :) 

Btw, no need to apologize, this is a toy project and any sort of question is welcome!

**Reply by joffrey-bion** [reply]
> How did you "know" which things to "extract" to kotlin from the javascript node module?

About this, I actually only extracted the things that I needed to use from Kotlin.
The actual npm package is what is used at runtime, so you don't need to declare anything you don't use directly (AFAIU).

So if I knew I needed -say- BlueprintJS's `Dialog` component, I would just go look at the props and Dialog component's Typescript definitions, and try to convert them to Kotlin externals manually.

The hardest part is to find how to express the actual React component class IMO (the props interfaces are quite straightforward, and the `RBuilder` extensions as well, they are just boilerplate). BlueprintJS uses vanilla React classes like `PureComponent` and the likes, but adapting them from Typescript was not obvious AFAIR (as they are not the same as the Kotlin wrappers).

**Comment by HoffiMuc** [comment]
well, I intended (after very brief first look arounds) maybe to use blueprintjs together with https://github.com/STRML/react-grid-layout

but integration of the latter just slips my abilities (I would understand if looking into this would not be a pleasure for you and you therefore just ignoring this :)

always getting a:

```
webpack-internal:///../../node_modules/react-grid-layout/build/GridItem.js:412 Uncaught TypeError: Cannot read property 'element' of undefined
    at eval (webpack-internal:///../../node_modules/react-grid-layout/build/GridItem.js:412)
    at Object.../../node_modules/react-grid-layout/build/GridItem.js (frontend.js:2630)
    at __webpack_require__ (frontend.js:30)
    at eval (webpack-internal:///../../node_modules/react-grid-layout/build/ReactGridLayout.js:17)
    at Object.../../node_modules/react-grid-layout/build/ReactGridLayout.js (frontend.js:2642)
    at __webpack_require__ (frontend.js:30)
    at eval (webpack-internal:///../../node_modules/react-grid-layout/index.js:1)
    at Object.../../node_modules/react-grid-layout/index.js (frontend.js:2701)
    at __webpack_require__ (frontend.js:30)```
```

kotlin wrapper try:

```kotlin
@file:JsModule("react-grid-layout") // import GridLayout from "react-grid-layout";
@file:JsNonModule
package net.strml.reactgridlayout

import react.*

//@JsModule("prop-types")

@JsName("default")
//external val GridLayout: RClass<ReactGridLayoutProps>
external val reactGridLayout : RClass<ReactGridLayoutProps>
@JsName("default")
external val GridItem: RClass<GridItemProps>

external interface ReactGridLayoutProps : RProps {
    var className: String
    var width: Int
    var autoSize: Boolean? // = true
    var cols: Int? // = 12
    var draggableCancel: String? // = ""
    var draggableHandle: String? // = ""
    var verticalCompact: Boolean? // = true
    var compactType: String // = "vertical"
    var layout: Array<Any>?
    var margin: Array<Int>? // = arrayOf(10, 10)
    var containerPadding: Array<Int>? // = margin
    var rowHeight: Int? // = 150
    var droppingItem: Any? // = object { var i: String; var w: Int; var h: Int }
    var isDraggable: Boolean? // = true
    var isResizable: Boolean? // = true
    var isBounded: Boolean? // = false
    var useCSSTransforms: Boolean? // = true
    var transformScale: Int? // = 1
    var preventCollision: Boolean? // = false
    var isDroppable: Boolean? // = false
    var resizeHandles: Array<String>? // = arrayOf("se")
    // resizeHandle?: ReactElement<any> | ((resizeHandleAxis: ResizeHandleAxis) => ReactElement<any>)
    var resizeHandle: ReactElement
    // onLayoutChange: (layout: Layout)  => void,
    var onLayoutChange:  (layouts: Array<Any>) -> Unit
    //
    // // Calls when drag starts.
    //     onDragStart: ItemCallback,
    // // Calls on each drag movement.
    //     onDrag: ItemCallback,
    // // Calls when drag is complete.
    //     onDragStop: ItemCallback,
    // // Calls when resize starts.
    //     onResizeStart: ItemCallback,
    // // Calls when resize movement happens.
    //     onResize: ItemCallback,
    // // Calls when resize is complete.
    //     onResizeStop: ItemCallback,
    // // Calls when an element has been dropped into the grid from outside.
    //     onDrop: (layout: Layout, item: ?LayoutItem, e: Event) // => void
    //
    // // Ref for getting a reference for the grid's wrapping div.
    // // You can use this instead of a regular ref and the deprecated `ReactDOM.findDOMNode()`` function.
    //     innerRef: ?React.Ref<"div">
}

external interface GridItemProps : RProps {
    var i: String
    var children: ReactElement

    // These are all in grid units, not pixels
    var x: Int
    var y: Int
    var w: Int
    var h: Int
    var minW: Int?
    var maxW: Int?
    var minH: Int?
    var maxH: Int?

    // If true, equal to `isDraggable: false, isResizable: false`.
    var static: Boolean?
    // If false, will not be draggable. Overrides `static`.
    var isDraggable: Boolean?
    // If false, will not be resizable. Overrides `static`.
    var isResizable: Boolean?
    // By default, a handle is only shown on the bottom-right (southeast) corner.
    // Note that resizing from the top or left is generally not intuitive.
    var resizeHandles: Array<String>?
    // If true and draggable, item will be moved only within grid.
    var isBounded: Boolean?
}
```

for

```javascript
// @flow
import React, {PropTypes} from 'react';
import {DraggableCore} from 'react-draggable';
import {Resizable} from 'react-resizable';
import {perc, setTopLeft, setTransform} from './utils';

import type {DragCallbackData, Position} from './utils';

type State = {
  resizing: ?{width: number, height: number},
  dragging: ?{top: number, left: number},
  className: string
};

/**
 * An individual item within a ReactGridLayout.
 */
export default class GridItem extends React.Component {

  static propTypes = {
    // Children must be only a single element
    children: PropTypes.element,

    // General grid attributes
    cols: PropTypes.number.isRequired,
    containerWidth: PropTypes.number.isRequired,
    rowHeight: PropTypes.number.isRequired,
    margin: PropTypes.array.isRequired,
    maxRows: PropTypes.number.isRequired,
    containerPadding: PropTypes.array.isRequired,

    // These are all in grid units
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    w: PropTypes.number.isRequired,
    h: PropTypes.number.isRequired,

    // All optional
    minW: function (props, propName, componentName) {
      const value = props[propName];
      if (typeof value !== 'number') return new Error('minWidth not Number');
      if (value > props.w || value > props.maxW) return new Error('minWidth larger than item width/maxWidth');
    },

    maxW: function (props, propName, componentName) {
      const value = props[propName];
      if (typeof value !== 'number') return new Error('maxWidth not Number');
      if (value < props.w || value < props.minW) return new Error('maxWidth smaller than item width/minWidth');
    },

    minH: function (props, propName, componentName) {
      const value = props[propName];
      if (typeof value !== 'number') return new Error('minHeight not Number');
      if (value > props.h || value > props.maxH) return new Error('minHeight larger than item height/maxHeight');
    },

    maxH: function (props, propName, componentName) {
      const value = props[propName];
      if (typeof value !== 'number') return new Error('maxHeight not Number');
      if (value < props.h || value < props.minH) return new Error('maxHeight smaller than item height/minHeight');
    },

    // ID is nice to have for callbacks
    i: PropTypes.string.isRequired,

    // Functions
    onDragStop: PropTypes.func,
    onDragStart: PropTypes.func,
    onDrag: PropTypes.func,
    onResizeStop: PropTypes.func,
    onResizeStart: PropTypes.func,
    onResize: PropTypes.func,

    // Flags
    isDraggable: PropTypes.bool.isRequired,
    isResizable: PropTypes.bool.isRequired,
    static: PropTypes.bool,

    // Use CSS transforms instead of top/left
    useCSSTransforms: PropTypes.bool.isRequired,

    // Others
    className: PropTypes.string,
    // Selector for draggable handle
    handle: PropTypes.string,
    // Selector for draggable cancel (see react-draggable)
    cancel: PropTypes.string
  };

  static defaultProps = {
    className: '',
    cancel: '',
    minH: 1,
    minW: 1,
    maxH: Infinity,
    maxW: Infinity
  };

  state: State = {
    resizing: null,
    dragging: null,
    className: ''
  };

  // Helper for generating column width
  calcColWidth(): number {
    const {margin, containerPadding, containerWidth, cols} = this.props;
    return (containerWidth - (margin[0] * (cols - 1)) - (containerPadding[0] * 2)) / cols;
  }

  /**
   * Return position on the page given an x, y, w, h.
   * left, top, width, height are all in pixels.
   * @param  {Number}  x             X coordinate in grid units.
   * @param  {Number}  y             Y coordinate in grid units.
   * @param  {Number}  w             W coordinate in grid units.
   * @param  {Number}  h             H coordinate in grid units.
   * @return {Object}                Object containing coords.
   */
  calcPosition(x: number, y: number, w: number, h: number, state: ?Object): Position {
    const {margin, containerPadding, rowHeight} = this.props;
    const colWidth = this.calcColWidth();

    const out = {
      left: Math.round((colWidth + margin[0]) * x + containerPadding[0]),
      top: Math.round((rowHeight + margin[1]) * y + containerPadding[1]),
      // 0 * Infinity === NaN, which causes problems with resize constriants;
      // Fix this if it occurs.
      // Note we do it here rather than later because Math.round(Infinity) causes deopt
      width: w === Infinity ? w : Math.round(colWidth * w + Math.max(0, w - 1) * margin[0]),
      height: h === Infinity ? h : Math.round(rowHeight * h + Math.max(0, h - 1) * margin[1])
    };

    if (state && state.resizing) {
      out.width = Math.round(state.resizing.width);
      out.height = Math.round(state.resizing.height);
    }

    if (state && state.dragging) {
      out.top = Math.round(state.dragging.top);
      out.left = Math.round(state.dragging.left);
    }

    return out;
  }

  /**
   * Translate x and y coordinates from pixels to grid units.
   * @param  {Number} top  Top position (relative to parent) in pixels.
   * @param  {Number} left Left position (relative to parent) in pixels.
   * @return {Object} x and y in grid units.
   */
  calcXY(top: number, left: number): {x: number, y: number} {
    const {margin, cols, rowHeight, w, h, maxRows} = this.props;
    const colWidth = this.calcColWidth();

    // left = colWidth * x + margin * (x + 1)
    // l = cx + m(x+1)
    // l = cx + mx + m
    // l - m = cx + mx
    // l - m = x(c + m)
    // (l - m) / (c + m) = x
    // x = (left - margin) / (coldWidth + margin)
    let x = Math.round((left - margin[0]) / (colWidth + margin[0]));
    let y = Math.round((top - margin[1]) / (rowHeight + margin[1]));

    // Capping
    x = Math.max(Math.min(x, cols - w), 0);
    y = Math.max(Math.min(y, maxRows - h), 0);

    return {x, y};
  }

  /**
   * Given a height and width in pixel values, calculate grid units.
   * @param  {Number} height Height in pixels.
   * @param  {Number} width  Width in pixels.
   * @return {Object} w, h as grid units.
   */
  calcWH({height, width}: {height: number, width: number}): {w: number, h: number} {
    const {margin, maxRows, cols, rowHeight, x, y} = this.props;
    const colWidth = this.calcColWidth();

    // width = colWidth * w - (margin * (w - 1))
    // ...
    // w = (width + margin) / (colWidth + margin)
    let w = Math.round((width + margin[0]) / (colWidth + margin[0]));
    let h = Math.round((height + margin[1]) / (rowHeight + margin[1]));

    // Capping
    w = Math.max(Math.min(w, cols - x), 0);
    h = Math.max(Math.min(h, maxRows - y), 0);
    return {w, h};
  }

  /**
   * This is where we set the grid item's absolute placement. It gets a little tricky because we want to do it
   * well when server rendering, and the only way to do that properly is to use percentage width/left because
   * we don't know exactly what the browser viewport is.
   * Unfortunately, CSS Transforms, which are great for performance, break in this instance because a percentage
   * left is relative to the item itself, not its container! So we cannot use them on the server rendering pass.
   *
   * @param  {Object} pos Position object with width, height, left, top.
   * @return {Object}     Style object.
   */
  createStyle(pos: Position): {[key: string]: ?string} {
    const {usePercentages, containerWidth, useCSSTransforms} = this.props;

    let style;
    // CSS Transforms support (default)
    if (useCSSTransforms) {
      style = setTransform(pos);
    }
    // top,left (slow)
    else {
      style = setTopLeft(pos);

      // This is used for server rendering.
      if (usePercentages) {
        style.left = perc(pos.left / containerWidth);
        style.width = perc(pos.width / containerWidth);
      }
    }

    return style;
  }

  /**
   * Mix a Draggable instance into a child.
   * @param  {Element} child    Child element.
   * @return {Element}          Child wrapped in Draggable.
   */
  mixinDraggable(child: React.Element<any>): React.Element<any> {
    return (
      <DraggableCore
        onStart={this.onDragHandler('onDragStart')}
        onDrag={this.onDragHandler('onDrag')}
        onStop={this.onDragHandler('onDragStop')}
        handle={this.props.handle}
        cancel={".react-resizable-handle" + (this.props.cancel ? "," + this.props.cancel : "")}>
        {child}
      </DraggableCore>
    );
  }

  /**
   * Mix a Resizable instance into a child.
   * @param  {Element} child    Child element.
   * @param  {Object} position  Position object (pixel values)
   * @return {Element}          Child wrapped in Resizable.
   */
  mixinResizable(child: React.Element<any>, position: Position): React.Element<any> {
    const {cols, x, minW, minH, maxW, maxH} = this.props;

    // This is the max possible width - doesn't go to infinity because of the width of the window
    const maxWidth = this.calcPosition(0, 0, cols - x, 0).width;

    // Calculate min/max constraints using our min & maxes
    const mins = this.calcPosition(0, 0, minW, minH);
    const maxes = this.calcPosition(0, 0, maxW, maxH);
    const minConstraints = [mins.width, mins.height];
    const maxConstraints = [Math.min(maxes.width, maxWidth), Math.min(maxes.height, Infinity)];
    return (
      <Resizable
        width={position.width}
        height={position.height}
        minConstraints={minConstraints}
        maxConstraints={maxConstraints}
        onResizeStop={this.onResizeHandler('onResizeStop')}
        onResizeStart={this.onResizeHandler('onResizeStart')}
        onResize={this.onResizeHandler('onResize')}>
        {child}
      </Resizable>
    );
  }

  /**
   * Wrapper around drag events to provide more useful data.
   * All drag events call the function with the given handler name,
   * with the signature (index, x, y).
   *
   * @param  {String} handlerName Handler name to wrap.
   * @return {Function}           Handler function.
   */
  onDragHandler(handlerName:string) {
    return (e:Event, {node, deltaX, deltaY}: DragCallbackData) => {
      if (!this.props[handlerName]) return;

      const newPosition: {top: number, left: number} = {top: 0, left: 0};

      // Get new XY
      switch (handlerName) {
        case 'onDragStart':
          // ToDo this wont work on nested parents
          const parentRect = node.offsetParent.getBoundingClientRect();
          const clientRect = node.getBoundingClientRect();
          newPosition.left = clientRect.left - parentRect.left;
          newPosition.top = clientRect.top - parentRect.top;
          this.setState({dragging: newPosition});
          break;
        case 'onDrag':
          if (!this.state.dragging) throw new Error('onDrag called before onDragStart.');
          newPosition.left = this.state.dragging.left + deltaX;
          newPosition.top = this.state.dragging.top + deltaY;
          this.setState({dragging: newPosition});
          break;
        case 'onDragStop':
          if (!this.state.dragging) throw new Error('onDragEnd called before onDragStart.');
          newPosition.left = this.state.dragging.left;
          newPosition.top = this.state.dragging.top;
          this.setState({dragging: null});
          break;
        default:
          throw new Error('onDragHandler called with unrecognized handlerName: ' + handlerName);
      }

      const {x, y} = this.calcXY(newPosition.top, newPosition.left);

      this.props[handlerName](this.props.i, x, y, {e, node, newPosition});
    };
  }

  /**
   * Wrapper around drag events to provide more useful data.
   * All drag events call the function with the given handler name,
   * with the signature (index, x, y).
   *
   * @param  {String} handlerName Handler name to wrap.
   * @return {Function}           Handler function.
   */
  onResizeHandler(handlerName: string) {
    return (e:Event, {node, size}: {node: HTMLElement, size: Position}) => {
      if (!this.props[handlerName]) return;
      const {cols, x, i, maxW, minW, maxH, minH} = this.props;

      // Get new XY
      let {w, h} = this.calcWH(size);

      // Cap w at numCols
      w = Math.min(w, cols - x);
      // Ensure w is at least 1
      w = Math.max(w, 1);

      // Min/max capping
      w = Math.max(Math.min(w, maxW), minW);
      h = Math.max(Math.min(h, maxH), minH);

      this.setState({resizing: handlerName === 'onResizeStop' ? null : size});

      this.props[handlerName](i, w, h, {e, node, size});
    };
  }

  render(): React.Element<any> {
    const {x, y, w, h, isDraggable, isResizable, useCSSTransforms} = this.props;

    const pos = this.calcPosition(x, y, w, h, this.state);
    const child = React.Children.only(this.props.children);

    // Create the child element. We clone the existing element but modify its className and style.
    let newChild = React.cloneElement(child, {
      // Munge a classname. Use passed in classnames and resizing.
      // React will merge the classNames.
      className: [
        'react-grid-item',
        child.props.className || '',
        this.props.className,
        this.props.static ? 'static' : '',
        this.state.resizing ? 'resizing' : '',
        this.state.dragging ? 'react-draggable-dragging' : '',
        useCSSTransforms ? 'cssTransforms' : ''
      ].join(' '),
      // We can set the width and height on the child, but unfortunately we can't set the position.
      style: {...this.props.style, ...child.props.style, ...this.createStyle(pos)}
    });

    // Resizable support. This is usually on but the user can toggle it off.
    if (isResizable) newChild = this.mixinResizable(newChild, pos);

    // Draggable support. This is always on, except for with placeholders.
    if (isDraggable) newChild = this.mixinDraggable(newChild);

    return newChild;
  }
}
```

**Reply by joffrey-bion** [reply]
Hey it's not a problem to ask, and it would be rude for me to just ignore your message ;)

Maybe the problem is how you declare your react component classes by just using;

```kotlin
@JsName("default")
external val reactGridLayout : RClass<ReactGridLayoutProps>
@JsName("default")
external val GridItem: RClass<GridItemProps>
```

I suspect several problems here:
- `reactGridLayout` and `GridItem` are declared as global *properties* (`val`) but I believe they should be actual *classes* instead
- both have `@JsName("default")`, doesn't that make them conflict at runtime? If I understood correctly, this means both declarations will have the same name in JS, I wonder how the Kotlin compiler even accepts it.

I wish I could help more, but to be honest I had a lot of trial and error until I got the blueprintjs externals to work properly, so I wouldn't be able to judge whether your declarations are correct by just looking at them (I wish I were that good 😄).

You should be able to get more help about this in Kotlin/JS [slack channel](https://kotlinlang.slack.com/archives/C0B8L3U69).

**Reply by HoffiMuc** [reply]
well the "default" seems to be just about the namespace ... js npm package components are imported to "default" in all the tutorials (not understanding this by myself :) )

it has something to do with js PropTypes/Flow static typing tool:

```javascript
GridItem.propTypes = {
  // Children must be only a single element
  children: _react.PropTypes.element,                             <----

  // General grid attributes
  cols: _react.PropTypes.number.isRequired,
  containerWidth: _react.PropTypes.number.isRequired,
  rowHeight: _react.PropTypes.number.isRequired,
  margin: _react.PropTypes.array.isRequired,
  maxRows: _react.PropTypes.number.isRequired,
  containerPadding: _react.PropTypes.array.isRequired,

  // These are all in grid units
  x: _react.PropTypes.number.isRequired,
  y: _react.PropTypes.number.isRequired,
  w: _react.PropTypes.number.isRequired,
  h: _react.PropTypes.number.isRequired,
  ...
```

**Reply by joffrey-bion** [reply]
> well the "default" seems to be just about the namespace ... js npm package components are imported to "default" in all the tutorials (not understanding this by myself :) )

Interesting. This is really not the way I understood `@JsName`. I couldn't find such examples of `@JsName` usage with duplicate values in the doc. The only one I found is in the [section about `@JsNonModule`](https://kotlinlang.org/docs/reference/js-modules.html#jsnonmodule-annotation):

```kotlin
@JsModule("hello")
@JsNonModule
@JsName("topLevelSayHello")
external fun sayHello(name: String)
```

And it really seems to be specific to the actual function declaration here, giving it a name for usage in JS.
Are you sure you're not confusing the use of `"default"` in `@JsModule` or things like that (which would IMO be better suited for namespaces).

Could you please point me to some of the tutorials using `@JsName("default")`? I'd like to understand this better, because if you're right I really need to update my understanding of this.

**Reply by HoffiMuc** [reply]
from: https://play.kotlinlang.org/hands-on/Building%20Web%20Applications%20with%20React%20and%20Kotlin%20JS/07_Using_Packages_From_NPM

```kotlin
@file:JsModule("react-player")
@file:JsNonModule

import react.*

@JsName("default")
external val reactPlayer: RClass<dynamic>
```

These last two lines are equivalent to a JavaScript import like `require("react-player").default;`. It tells the compiler that we're certain we'll get a component conforming to RClass<dynamic> at runtime.

**Reply by joffrey-bion** [reply]
About your actual error:

```
webpack-internal:///../../node_modules/react-grid-layout/build/GridItem.js:412 Uncaught TypeError: Cannot read property 'element' of undefined
    at eval (webpack-internal:///../../node_modules/react-grid-layout/build/GridItem.js:412)
```

Maybe you need to manually add an `npm` dependency on the [prop-types](https://www.npmjs.com/package/prop-types) module?
I guess this is the one that's supposed to bring the `_react.PropTypes.element` value, and it is [declared in the package.json of the library](https://github.com/STRML/react-grid-layout/blob/master/package.json#L40).
It should be automatically brought (transitively) but I don't trust transitive deps anymore, I had to add them in a couple situations, and still don't get why.

**Reply by joffrey-bion** [reply]
> These last two lines are equivalent to a JavaScript import like require("react-player").default;. It tells the compiler that we're certain we'll get a component conforming to RClass at runtime.

Thanks a lot! I will look at the tutorial again then. I guess the way `@JsName` works is different for externals and for regular Kotlin/JS declarations. The former is to tell the compiler that we know a declaration exists in JS under that name, and the latter is to tell the compiler to rename the element when converting to JS.

**Reply by HoffiMuc** [reply]
well importing npm prop-types didn't solve the problem ...

as _react.PropTypes.element is inside the implementation of the external class/function GridItem

and seems to be a static member referring to something

**Comment by HoffiMuc** [comment]
well I am continuing my "journey" to kotlinjs with blueprintjs ... also dived a bit deeper into dukat. also having all React Types in wrapped Kotlinjs (by dukat).

now trying to get @blueprintjs/select module things to work ...

so if you fancy ...

trying to wrap blueprintjs to kotlinjs ... @blueprintjs/core package worked fine so far, but I got stuck on the select package.
I used dukat to generate all the "wrapping" parts and now try to use the select component of blueprintjs following https://codesandbox.io/s/blueprint-sandbox-et9xy?file=/src/SelectExample.tsx

```
  public render() {
    const buttonText = this.state.film.title;
    return (
      <Example header="Select Sandbox">
        <FilmSelect
          items={Films.TOP_100_FILMS}
          itemPredicate={Films.filterFilm}
          itemRenderer={Films.renderFilm}
          noResults={<MenuItem disabled={true} text="No results." />}
          onItemSelect={this.handleValueChange}
        >
          <Button text={buttonText} rightIcon="caret-down" />
        </FilmSelect>
      </Example>
    );
  }
```

my try is:

```
val FilmSelect = Select.ofType<Film>()
val exampleSelect = functionalComponent <SelectExampleProps> { props ->
        val (selectExampleState, setSelectExampleState) = useState(SelectExampleState(TOP_100_FILMS[0]))
        child(genericCard("Select Sandbox")) {
            child(FilmSelect::class) {
                attrs.items =  TOP_100_FILMS
                attrs.itemPredicate = filterFilm
                attrs.itemRenderer = renderFilm
                attrs.noResults = noResultsFunc
                attrs.onItemSelect = handleValueChange
                child(Button::class) {
                    attrs.text = selectExampleState.selectedFilm.title
                    attrs.rightIcon = "caret-down"
                }
            }
        }
    val handleValueChange = { film: Film, event: SyntheticEvent__1<HTMLElement> ->
        setSelectExampleState(SelectExampleState(film))
    }
}
```

so itemRenderer is

`typealias ItemRenderer<T> = (item: T, itemProps: IItemRendererProps) -> ReactElement?`

how can I implement such a funtion returning a ReactElement but not being "in scope" ??

my try so far is:

```
val renderFilm: ItemRenderer<Film> = { film: Film, itemProps: IItemRendererProps ->
    if (!itemProps.modifiers.matchesPredicate) {
        null;
    }
    val text = "${film.rank}. ${film.title}"
    child(MenuItem::class) {
        attrs.key = film.rank
        attrs.active = itemProps.modifiers.active
        attrs.disabled = itemProps.modifiers.disabled
        attrs.label = film.year.toString()
        attrs.onClick = itemProps.handleClick
        attrs.text = highlightText(text, itemProps.query)
    }
}
val filterFilm : ItemPredicate<Film> = { query: String, film: Film, index: Number, exactMatch: Boolean ->
    "${film.rank}. ${film.title.toLowerCase()} ${film.year}".indexOf(query.toLowerCase()) >= 0
}
```

but obviously the lambda has no child(MenuItem::class) method I can use ...
how would I implement an ItemRenderer which returns the ReactElement ??

If I bother you too much, just say a word :)

**Reply by joffrey-bion** [reply]
Hey, thanks for coming back with news!

> also having all React Types in wrapped Kotlinjs (by dukat)

Do you mean Dukat automatically uses kotlin-react wrapper types for Kotlin types when running on blueprintjs?
Or do mean you generated kotlin react types yourself from `react`'s types using Dukat?

> @blueprintjs/core package worked fine so far

Do you mean running Dukat on `@blueprintjs/core` worked fine? Did you need to make adjustments?

Or did you write the external declarations completely manually like I did?

> I used dukat to generate all the "wrapping" parts

Does the generated code compile fine? Or did you need to make adjustments?

> how can I implement such a funtion returning a ReactElement but not being "in scope" ??

It took me a while to figure that out, but there is a `buildElement` top-level function (provided by the `kotlin-react` module I believe).
I make use of it here for instance:
https://github.com/joffrey-bion/seven-wonders/blob/fc24ce2fb865f44d78e9e7e2519b3a7b19d4b8d5/sw-ui/src/main/kotlin/org/luxons/sevenwonders/ui/components/lobby/RadialPlayerList.kt#L22

It's a bad example because an `RBuilder` is already in scope, but it doesn't use that one of course.
Here is another example:
https://github.com/joffrey-bion/seven-wonders/blob/fc24ce2fb865f44d78e9e7e2519b3a7b19d4b8d5/sw-ui/src/main/kotlin/org/luxons/sevenwonders/ui/components/home/ChooseNameForm.kt#L47

I think you should be able to do this:
```kotlin
val renderFilm: ItemRenderer<Film> = { film: Film, itemProps: IItemRendererProps ->
    if (!itemProps.modifiers.matchesPredicate) {
        null;
    }
    val text = "${film.rank}. ${film.title}"
    buildElement {
        child(MenuItem::class) {
            attrs.key = film.rank
            attrs.active = itemProps.modifiers.active
            attrs.disabled = itemProps.modifiers.disabled
            attrs.label = film.year.toString()
            attrs.onClick = itemProps.handleClick
            attrs.text = highlightText(text, itemProps.query)
        }
    }
}
```

> If I bother you too much, just say a word :)

No problem at all, I may have some delays in my answers, but apart from that I'm pretty happy to discuss things.

**Reply by HoffiMuc** [reply]
well not sure what I am really doing, but it is about this:

```
cd build/js/node_modules/@blueprintjs
find core/lib/esm -type f -name 'index.d.ts' | xargs dukat -d dukatcore
find select/lib/esm/ -type f -name 'index.d.ts' | xargs dukat -d dukatselect
```

and I also cloned https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react
to get all the react types to kotlin types

```
cd DefinitelyTyped/types/react/
dukat -d dukat *.d.ts
```

from the latter one I copied `index.module_react.kt`, `index_React.module_react.kt` and `nonDeclarations.React.kt` to my project.

After a lot of fixing and copying typealiases I got the copied over (similar to yours) stuff from the bluprintjs dukat kotlin to run.

Having the kotlin react types from DefinitelyTyped in the project, adjustments should be way less than in your more manual approach.

But still there was a lot of work to find things with the `https://github.com/ggreer/the_silver_searcher` that I wanted either to copy over or else to adjust on the dukat kotlin files...

atm, I am stuck with getting the `Select Sandbox` https://codesandbox.io/s/blueprint-sandbox-et9xy?file=/src/SelectExample.tsx thing to get running.

Whole day yesterday took me to get it compiled ... now I get runtime errors that come deep down of react and I wonder how to correlate them to my code at all.

If you fancy you can have a look at my project as it is just a "stub" project anyway ...

https://github.com/HoffiMuc/web-app-react-kotlin-js-gradle  (select component is commented out in Main.kt so that the frontend comes up without an error on startup)

**Reply by joffrey-bion** [reply]
Very interesting! I don't think you have to clone the `DefinitelyTyped` repo. You should be able to just declare npm dependencies on the types of a specific module in your gradle file, like `npm("@types/react")` and `npm("@types/react-dom")`, and this will only fetch the ones you need from DefinitelyTyped.
Maybe that's what's wrong with my approach, though. I can't say I have succeeded yet, so take this with a grain of salt.

Actually, what you did is similar to what I've been trying to automate in https://github.com/joffrey-bion/kotlin-blueprintjs.

Currently what the `buildSrc` contains is a custom Gradle task that will do a couple things:

1. go to the module of your choice (in `build/js/node_modules`, so like you did)
2. find where the TypeScript index is by reading the `package.json` (which in the case of `@blueprintjs/core` yields `lib/esm/index.d.ts`, just like you found)
3. Run dukat on this TypeScript file, specifying the chosen output folder (by default `src/main/generated`, but I intend to also nest under a folder that matches the package)
4. Perform search&replace on the contents of the generated Kotlin files to fix some of Dukat's mistakes

This custom Gradle task is then used in the main project's `build.gradle.kts` to process the `@blueprintjs/core` module (but could very well be used on any module I believe, so in particular maybe `@blueprintjs/select` that you need).

In the README of that project, I'm trying to keep a list of the manual changes that I still need to perform to get it to compile, but it's not complete yet (it doesn't compile after just these changes, it needs more tweaks).
I'd like to automate as much as possible in order to update smoothly the react and blueprintjs versions over time. The more manual work I do and check in git, the more needs to be manually checked again against new versions.

One of the things I need to get rid of is the typealiases stuff, just like you said I think. I'm using `@JsModule` so I can't put typealiases in the same file as the externals, which is pretty annoying.

> If you fancy you can have a look at my project as it is just a "stub" project anyway

Thanks for the link! I'm not sure I will have time in the coming days to run it and/or see what's wrong at runtime, unfortunately. For runtime errors it's never easy I believe, you have to keep in mind that external declarations can be anything, nothing guarantees it matches the actual JS code. Therefore, the usual assumptions about what is correct when the code compiles don't really hold anymore (e.g. what properties are present on an object, what functions you can call etc.).
